gui/designwidget.cc

```diff
@@ -265,11 +265,49 @@ void DesignWidget::newContext(Context *ctx)
         std::lock_guard<std::mutex> lock_ui(ctx->ui_mutex);
         std::lock_guard<std::mutex> lock(ctx->mutex);
         
-        getTreeByElementType(ElementType::BEL)->loadContext(ElementType::BEL, ctx);
-        getTreeByElementType(ElementType::WIRE)->loadContext(ElementType::WIRE, ctx);
-        getTreeByElementType(ElementType::PIP)->loadContext(ElementType::PIP, ctx);
-        getTreeByElementType(ElementType::CELL)->loadContext(ElementType::CELL, ctx);
-        getTreeByElementType(ElementType::NET)->loadContext(ElementType::NET, ctx);
+        {
+            std::map<std::pair<int, int>, std::vector<BelId>> belMap;
+            for (auto bel : ctx->getBels()) {
+                auto loc = ctx->getBelLocation(bel);
+                belMap[std::pair<int, int>(loc.x, loc.y)].push_back(bel);
+            }
+            auto belGetter = [](Context *ctx, BelId id) { return ctx->getBelName(id); };
+            
+            getTreeByElementType(ElementType::BEL)->loadData(std::unique_ptr<TreeModel::ElementXYRoot<BelId>>(
+                    new TreeModel::ElementXYRoot<BelId>(ctx, belMap, belGetter, ElementType::BEL)));
+        }
+
+#ifdef ARCH_ICE40        

```

Comment:
I'd rather have a proper interface for architecture-specific UI implementation, but I understand that this would involve refactoring.